meorg_client/client.py

```diff
@@ -25,84 +26,128 @@ def __init__(self, base_url=mcc.MEORG_URL, email=None, password=None):
         dict or requests.Response
-            Decoded JSON response, or raw response.
+            Dictionary or Request object, depending on context.
 
         Raises
         ------
-        Exception
-            Raised when there is an issue.
+        mx.InvalidHTTPMethodException
+            Raised when the specified method is invalid.
+        RequestException
+            Raised when the request fails.
         """
 
-        # Cast to uppercase, for consisency
         method = method.upper()
 
+        # Check that the method is allowed.
         if method not in mcc.VALID_METHODS:
-            raise Exception(f"Invalid method {method}")
+            raise mx.InvalidHTTPMethodException(method)
 
-        # GET/PUT requests have the data interpolated into the url
-        if method in mcc.INTERPOLATING_METHODS:
-            endpoint = endpoint.format(**data)
+        # Get the function and URL
+        func = getattr(requests, method.lower())
+        url = self._get_url(endpoint, **url_params)
 
-        url = urljoin(self.base_url, endpoint)
-        all_headers = {**self.headers, **headers} if headers else self.headers
+        # Assemble the headers
+        _headers = self._merge_headers(headers)
 
-        # Update the client with the last response
-        self.last_response = requests.request(
-            method, url, data=data, headers=all_headers
+        # Make the request, set it as the last response for future use
+        self.last_response = func(
+            url, data=data, json=json, headers=_headers, files=files, **kwargs
         )
 
-        # RFC 2616 states status in the 2xx range are considered successful
-        if self.last_response.status_code in mcc.HTTP_STATUS_SUCCESS_RANGE:
-            # Return JSON if that's what it is (this should be the default)
-            if (
-                self.last_response.headers.get("Content-Type", str)
-                == mcc.HTTP_CONTENT_TYPE_JSON
-                and return_json == True
-            ):
-                return self.last_response.json()
-
-            return self.last_response
-
-        else:
+        # Check to see if it was successful
+        if self.last_response.status_code not in mcc.HTTP_STATUS_SUCCESS_RANGE:
             raise RequestException(
                 self.last_response.status_code, self.last_response.text
             )
 
+        # This is the default
+        if return_json:
+            return self.last_response.json()
+
+        # For flexibility
+        return self.last_response
+
+    def _get_url(self, endpoint, **kwargs):
+        """Get the well-formed URL for the call.
+
+        Parameters
+        ----------
+        endpoint : str
+            Endpoint to be appended to the base URL.
+        **kwargs :
+            Key/value pairs to interpolate into the URL template.
+
+        Returns
+        -------
+        str
+            URL.
+        """
+        return urljoin(self.base_url + "/", endpoint).format(**kwargs)
```

Comment:
Considering that `base_url` should have `/` from 34, not sure adding the / would help

Reply:
See comment above.

Reply:
This is also just a hammer solution that will force the slash if for some reason it has been omitted somewhere along the way. `urljoin()` has some smarts in it to ignore if not needed.

Reply:
Maybe a comment would be suitable here